docs/src/proposals/traced-transaction-construction.md

```diff
@@ -0,0 +1,247 @@
+# Traced transaction construction via RPC's `simulateTransaction`
+
+## Problem
+
+With more composition comes in the Solana's program ecosystem, it will be harder
+and harder for the higher-level consuming program to correctly construct
+lower-level program's instructions by precisely specifying each and every
+accessed accounts with the `read_only` or `read_write` flag.
+
+Also, there will be no way to signal the need of additional accounts for the
+arbitrarily customized behavior from the lower-level program to the higher-level
+one under the standardized instruction API like spl-token, even though we're
+about to explore the flexibility of allowing customzed programs conforming to a
+standard like the itoken.
+
+These preceding runtime limitations are applicable both to on-chain (instruction
+construction for CPI) or off-chain (transaction construction by end-user
+interfaces) boundaries likewise.
+
+Lastly, it's practically impossible to change those accessed account addresses
+once deployed while retaining compatibility, given the fact the consumers are
+expected to hard-code the account addresses for consumption, even if the
+deployed programs can be now upgradeable. This severely hampers ongoing
+consumed-program's implementation refinement in the areas of data migration and
+improved concurrency, by preventing transparent account switching and
+sharding with vectorized accounts, respectively.
```

Comment:
what is meant by "vectorized accounts"?

Reply:
@jstarry don't be bothered with it.. ;) well, I used this fancy adjective, remotely referencing the usage like `vectorization with SSE4/AVX2`. Maybe `sharding` by itself should be enough.

`Sharding` here means avoiding write lock to the same internal account by using different accounts to write by deriving from the first byte of other account addresses like order_id for DEX. So, program can avoid write lock contention easily. Also, see the sharding example for code.

I want to say, hiding internal account addresses in this way makes these optimizations can be introduced by drop-in basis.

Reply:
Ah got it, I like the idea of abstracting away implementation details like this, thanks for the explanation!